### ISSUE TITLE
collide step: group_name with spaces

### ISSUE THREAD

**Issue by andreagalle** [Toontown-Event-Horizon/panda-utils#1]
In my .egg file I have a Group name with spaces. The collide step doesn't find any group there nor the rename step is able to rename it without spaces. Is there the possibility to support `group_name` with whitespaces too?

    collide:
      - group_name: "Tempio Maggiore di Roma"
        method: polyset
        flags: keep,descend

**Comment by Wizzerinus** [comment]
Hi!

I will need the following information for debugging:

1) The egg file the step operates on. Since you're using the pipeline, it should be available in the intermediate folder. If you do not want to/are unable to provide the full file, you can cut out the unrelated sections.
2) The pipeline command you run. If you use composer, it can be obtained by setting the environmental variable `PANDA_UTILS_LOGGING=1`.
3) The version of Panda Utils used (can be obtained by using `pip freeze` or similar commands).

(I really should add an issue template with these)

**Comment by andreagalle** [comment]
I am going to give you anything you need :) 

1. These are the .obj/.blend/.bam/.egg files within the intermediate directory, along with the model-config.yml file to reproduce the pipeline:
[misc.zip](https://github.com/Toontown-Event-Horizon/panda-utils/files/13266981/misc.zip)
2. This is the command I ran along with its output:
        (panda3D) > python -m panda_utils.assetpipeline input/asset-folder phase_1/models/tiberriver phase_1/maps preblend blend2bam bam2egg group_rename[] collide[] egg2bam
        panda_utils.pipeline: INFO     Running Pipeline with parameters: ['preblend', 'blend2bam', 'bam2egg', 'group_rename[]', 'collide[]', 'egg2bam']
        panda_utils.pipeline.models: INFO     Asset Folder: Converting to blend: test4_buildings.obj, test4_buildings.obj.mtl, test4_buildings.obj_textures, test4_highways.obj, test4_highways.obj.mtl, test4_highways.obj_textures
        DEBUG:BlenderGIS-master.core.checkdeps:GDAL Python binding unavailable
        DEBUG:BlenderGIS-master.core.checkdeps:PyProj available
        DEBUG:BlenderGIS-master.core.checkdeps:Pillow available
        DEBUG:BlenderGIS-master.core.checkdeps:ImageIO Freeimage plugin available
        panda_utils.pipeline.models: INFO     Asset Folder: Exporting to GLTF: asset-folder.blend
        DEBUG:BlenderGIS-master.core.checkdeps:GDAL Python binding unavailable
        DEBUG:BlenderGIS-master.core.checkdeps:PyProj available
        DEBUG:BlenderGIS-master.core.checkdeps:Pillow available
        DEBUG:BlenderGIS-master.core.checkdeps:ImageIO Freeimage plugin available
        panda_utils.pipeline.models: INFO     Asset Folder: Converting to bam: asset-folder.glb
        panda_utils.pipeline.models: INFO     Asset Folder: Converting asset-folder.bam from Bam to Egg
        panda_utils.converter: INFO     Recompilation not needed!
        panda_utils.pipeline.models: INFO     Asset Folder: Adding collisions to asset-folder.egg/Sinagoga: flags=keep descend method=Polyset
        panda_utils.pipeline.models: INFO     Asset Folder: Adding collisions to asset-folder.egg/Building0: flags=keep descend method=Polyset
        panda_utils.pipeline.models: INFO     Found the named group!
        panda_utils.pipeline.models: INFO     Asset Folder: Adding collisions to asset-folder.egg/Building1: flags=keep descend method=Polyset
        panda_utils.pipeline.models: INFO     Found the named group!
        panda_utils.pipeline.models: INFO     Asset Folder: Adding collisions to asset-folder.egg/Building2: flags=keep descend method=Polyset
        lding5: flags=keep descend method=Polyset
        panda_utils.pipeline.models: INFO     Found the named group!                                 lding3: flags=keep descend method=Polyset
        panda_utils.pipeline.models: INFO     Asset Folder: Adding collisions to asset-folder.egg/Building6: flags=keep descend method=Polyset                                                    lding4: flags=keep descend method=Polyset
        panda_utils.pipeline.models: INFO     Found the named group!
        panda_utils.pipeline.models: INFO     Asset Folder: Adding collisions to asset-folder.egg/Building5: flags=keep descend method=Polysetlding7: flags=keep descend method=Polyset
        panda_utils.pipeline.models: INFO     Found the named group!                                 lding6: flags=keep descend method=Polyset
        panda_utils.pipeline.models: INFO     Asset Folder: Adding collisions to asset-folder.egg/Building8: flags=keep descend method=Polyset                                                    lding7: flags=keep descend method=Polyset
        panda_utils.pipeline.models: INFO     Found the named group!
        panda_utils.pipeline.models: INFO     Asset Folder: Adding collisions to asset-folder.egg/Building8: flags=keep descend method=Polysetlding9: flags=keep descend method=Polyset
        panda_utils.pipeline.models: INFO     Found the named group!                                 lding9: flags=keep descend method=Polyset
        panda_utils.pipeline.models: INFO     Asset Folder: Copying asset-folder.egg into the dist directory                                                                                      rectory
        panda_utils.pipeline.models: INFO     Asset Folder: Converting asset-folder.egg to bam  
3. this are the pkgs installed in my venv:
(panda3D) > pip freeze                                                                                         lding9: flags=keep descend method=Polyset
numpy==1.24.4
panda-utils==1.5.3                                                                           rectory
Panda3D==1.10.12
panda3d-blend2bam==0.23.0
panda3d-gltf==1.0.1
panda3d-simplepbr==0.10
PyYAML==6.0.1

**Comment by andreagalle** [comment]
I am going to give you anything you need :) 

1. These are the .obj/.blend/.bam/.egg files within the intermediate directory, along with the model-config.yml file to reproduce the pipeline: [misc.zip](https://github.com/Toontown-Event-Horizon/panda-utils/files/13266981/misc.zip)

2. This is the command I ran along with its output:

```
(panda3D) PS > python -m panda_utils.assetpipeline input/asset-folder phase_1/models blend2bam bam2egg group_rename[] collide[] egg2bam
panda_utils.pipeline: INFO     Running Pipeline with parameters: ['preblend', 'blend2bam', 'bam2egg', 'group_rename[]', 'collide[]', 'egg2bam']
panda_utils.pipeline.models: INFO     Asset Folder: Converting to blend: test4_buildings.obj, test4_buildings.obj.mtl, test4_buildings.obj_textures, test4_htl, test4_highways.obj_textures
DEBUG:BlenderGIS-master.core.checkdeps:GDAL Python binding unavailable
DEBUG:BlenderGIS-master.core.checkdeps:PyProj available
DEBUG:BlenderGIS-master.core.checkdeps:Pillow available
DEBUG:BlenderGIS-master.core.checkdeps:ImageIO Freeimage plugin available
panda_utils.pipeline.models: INFO     Asset Folder: Exporting to GLTF: asset-folder.blend
DEBUG:BlenderGIS-master.core.checkdeps:GDAL Python binding unavailable
DEBUG:BlenderGIS-master.core.checkdeps:PyProj available
DEBUG:BlenderGIS-master.core.checkdeps:Pillow available
DEBUG:BlenderGIS-master.core.checkdeps:ImageIO Freeimage plugin available
panda_utils.pipeline.models: INFO     Asset Folder: Converting to bam: asset-folder.glb
panda_utils.pipeline.models: INFO     Asset Folder: Converting asset-folder.bam from Bam to Egg
panda_utils.converter: INFO     Recompilation not needed!
panda_utils.pipeline.models: INFO     Asset Folder: Adding collisions to asset-folder.egg/Sinagoga: flags=keep descend method=Polyset
panda_utils.pipeline.models: INFO     Asset Folder: Adding collisions to asset-folder.egg/Building0: flags=keep descend method=Polyset
panda_utils.pipeline.models: INFO     Found the named group!
panda_utils.pipeline.models: INFO     Asset Folder: Adding collisions to asset-folder.egg/Building1: flags=keep descend method=Polyset
panda_utils.pipeline.models: INFO     Found the named group!
panda_utils.pipeline.models: INFO     Asset Folder: Adding collisions to asset-folder.egg/Building2: flags=keep descend method=Polyset
panda_utils.pipeline.models: INFO     Asset Folder: Adding collisions to asset-folder.egg/Building6: flags=keep descend method=Polyset                                                    lding3: flags=keep descend method=Polyset
panda_utils.pipeline.models: INFO     Found the named group!
panda_utils.pipeline.models: INFO     Asset Folder: Adding collisions to asset-folder.egg/Building4: flags=keep descend method=Polysetlding7: flags=keep descend method=Polyset
panda_utils.pipeline.models: INFO     Found the named group!                                 lding5: flags=keep descend method=Polyset
panda_utils.pipeline.models: INFO     Asset Folder: Adding collisions to asset-folder.egg/Building9: flags=keep descend method=Polyset
panda_utils.pipeline.models: INFO     Found the named group!
panda_utils.pipeline.models: INFO     Asset Folder: Copying asset-folder.egg into the dist directory
panda_utils.pipeline.models: INFO     Asset Folder: Converting asset-folder.egg to bam
```

3. this are the pkgs installed in my venv:

```
(panda3D) > pip freeze
numpy==1.24.4
panda-utils==1.5.3
Panda3D==1.10.12
panda3d-blend2bam==0.23.0
panda3d-gltf==1.0.1
panda3d-simplepbr==0.10
PyYAML==6.0.1
```

**Comment by Wizzerinus** [comment]
So there are two issues here:
- You're looking for `Tempio Maggiore di Roma`, but the node name is actually `Building Tempio Maggiore di Roma`:

![image](https://github.com/Toontown-Event-Horizon/panda-utils/assets/52964431/a0e2cb47-8735-4eb4-8131-cd0e0988f5e2)

- However, Panda Utils does not properly purge and reapply the quoting from the node names. This will be fixed in version 1.6 (which is currently in development, and I will be pushing it most likely later today).

- There are also multiple nodes with that name, which may be problematic for the collide step. I will check out the file's structure in Blender a bit later due to it being a potential incompatibility (I suspect it won't be, but a check can never be wrong anyway).

**Comment by andreagalle** [comment]
Of course if I skip the `group_rename` (here below):

    group_rename:
      Building Tempio Maggiore di Roma: Sinagoga
    collide:
      - group_name: Sinagoga
        method: polyset
        flags: keep,descend

the `collide` step fails as well on the "Building Tempio Maggiore di Roma" group (with whitespaces in its name).

    collide:
      - group_name: Building Tempio Maggiore di Roma
        method: polyset
        flags: keep,descend

**Comment by andreagalle** [comment]
> So there are two issues here:
> 
> * You're looking for `Tempio Maggiore di Roma`, but the node name is actually `Building Tempio Maggiore di Roma`:
> 
> ![image](https://user-images.githubusercontent.com/52964431/280726347-a0e2cb47-8735-4eb4-8131-cd0e0988f5e2.png)
> 
> * However, Panda Utils does not properly purge and reapply the quoting from the node names. This will be fixed in version 1.6 (which is currently in development, and I will be pushing it most likely later today).
> * There are also multiple nodes with that name, which may be problematic for the collide step. I will check out the file's structure in Blender a bit later due to it being a potential incompatibility (I suspect it won't be, but a check can never be wrong anyway).

How did you enabled such log level? This is diferent from my output. (I think I was really looking for "Building Tempio Maggiore di Roma") isn't it?

**Comment by Wizzerinus** [comment]
Yeah this is because of the issue with quotes (I'm fixing it as we talk). YAML automatically unquotes any quoted string so `value: Building Tempio Maggiore di Roma` and `value: "Building Tempio Maggiore di Roma"` are exactly the same.

My output was from python console running eggtree parser directly, not from the pipeline. But apologies, I see you realized that according to the attached model-config file. So it's just the quote bug that remains.

**Comment by Wizzerinus** [comment]
The potential fix should be now available as a part of the 1.6b3 release (`pip install panda-utils==1.6b3`). Note that if you rely on fully-transparent vertex coloring in your models (which is extremely unlikely but I'm mentioning it anyway) you will have to make adjustments to your config file.

Since there is currently no tests for the eggtree stuff (I plan to add them when I'm less busy) I'll wait on closing until you confirm the fix works properly.

**Comment by andreagalle** [comment]
Since you are already putting your hands in it, I was wondering if the `collide` step could support **fnmatch syntax** too, as the `group_remove` one does, since I need to apply it to every group within an `.egg` file (made of buildings solely). Is that possible?

**Comment by andreagalle** [comment]
> * looking for `Tempio Maggiore di Roma`, but the node name is actually `Building Tempio Maggiore di Roma`

I don't know why am I looking for "Tempio Maggiore di Roma", rather than "Building Tempio Maggiore di Roma"

**Comment by Wizzerinus** [comment]
I derived that from the model-config snippet provided in the issue description, the actual model-config file has the correct name so I think it should work.

As for fnmatch syntax, I can probably add that to a bunch of other commands as well.

**Comment by Wizzerinus** [comment]
Apparently there is a bug in the 1.6b3 version that anything that uses EggString (which includes basically every model since things like UV also use it) broke due to the change. I will yank that version and make a new one.

**Comment by Wizzerinus** [comment]
Alright 1.6b4 should be ready for use, let me know if you encounter any other issues.

**Comment by andreagalle** [comment]
That's wonderful: decompiling to `.egg`, the resulting `.bam` file, I get the following:

```
<Group> {
  <Transform> {
    <Rotate> { 90.000001961151 1 0 0 }
  }
  <Group> Sinagoga {
    <Collide> { Polyset descend }
    <VertexPool> vpool-collision.vpool1 {
      <Vertex> 0 {
        259.311004638672 110.106994628906 26.0000019073486
        <Normal> { 0.919978022575378 -0.391970068216324 -1.71335550191998e-08 }
      }
      .................
```

this means you fixed both issues for group name with whitespaces, in the `group_rename` and `collide` steps!

Now, since I am waiting for you to release the **fnmatch** syntax support (💯), I will in the meanwhile develop some sort of `.egg` file parser to retrieve all the buildings within my `.egg` files and produce a corresponding `model-config.yml` with them all!

**Comment by Wizzerinus** [comment]
fnmatch is implemented in 1.6b4 already.
